src/github.com/couchbase/sync_gateway/rest/handler.go

```diff
@@ -156,8 +170,29 @@ func (h *handler) invoke(method handlerMethod) error {
 		if err != nil {
 			return err
 		}
+
+		//If the Database is currently online process as normal
+		dbState := atomic.LoadUint32(&dbContext.State)
+		if (dbState == db.DBOnline) {
+			//Take a readlock for any method that does not run offline, taking a DB offline will
+			//block until all normal calls have returned and released their read locks.
+			if (!h.runOffline) {
+				//get a read lock on the dbContext
+				dbContext.AccessLock.RLock()
```

Comment:
It feels like there's a potential race condition here, in a scenario like:
1. This code completes the evaluation on line 176.
2. TakeDbOffline is called, updates state to stopping and blocks waiting for the write lock
3. This code gets to line 181, and blocks here trying to get the RLock 
4. After the write lock is released, this request proceeds, even though the DB is offline.